controllers/machinehealthcheck_controller.go

```diff
@@ -75,8 +89,19 @@ func (r *MachineHealthCheckReconciler) SetupWithManager(mgr ctrl.Manager, option
 		return errors.Wrap(err, "error setting index fields")
 	}
 
+	// Add index to Machine for listing by Node reference
+	if err := mgr.GetCache().IndexField(&clusterv1.Machine{},
+		machineNodeNameIndex,
+		r.indexMachineByNodeName,
+	); err != nil {
+		return errors.Wrap(err, "error setting index fields")
+	}
+
 	r.controller = controller
 	r.recorder = mgr.GetEventRecorderFor("machinehealthcheck-controller")
+	r.scheme = mgr.GetScheme()
+	r.clusterNodeInformers = make(map[client.ObjectKey]cache.Informer)
+	r.clusterNodeInformersLock = sync.RWMutex{}
```

Comment:
nit (non-blocking): the zero value of a RWMutex is ready for use, so this is not strictly necessary